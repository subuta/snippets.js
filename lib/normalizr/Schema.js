import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

import parseRelations from 'lib/_utils/parseRelations'

export default (Models = []) => {
  // omit junction table.
  Models = _.omitBy(Models, { isJunction: true })

  const schemaDefs = _.flatten(_.map(Models, ({schema}) => {
    const {tableName} = schema
    const model = pluralize.singular(_.camelCase(tableName))

    return [
      `export const ${model} = new schema.Entity('${model}')`,
      `export const ${model}List = new schema.Array(${model})`,
      s.EOL
    ]
  }))

  const associations = _.flatten(_.map(Models, (Model) => {
    const {schema} = Model
    const {tableName} = schema
    const model = pluralize.singular(_.camelCase(tableName))

    const relations = _.transform(parseRelations(Model), (result, value, key) => {
      result[key] = pluralize.isPlural(value) ? `[${pluralize.singular(value)}]` : value
    }, {})

    return build`
      ${model}.define(${s.raw(relations)})
    ` + s.EOL
  }))

  let models = _.flatten(_.map(_.keys(Models), (m) => {
    const model = pluralize.singular(m)
    return [
      model,
      `${model}List`
    ]
  }))
  models = _.zipObject(models, models)

  const imports = s.import([
    ['lodash', '_'],
    ['pluralize', 'pluralize'],
    ['src/modules', null, [
      ['fetchEntities']
    ]]
  ])

  return build`
    ${imports}
    import * as normalizr from 'normalizr'
    
    const { schema } = normalizr
    
    ${schemaDefs}
    
    ${associations}
    
    const models = ${s.raw(models)}
    
    // denormalize data using schema.
    export const denormalize = (data, modelName, state) => {
      const schema = models[pluralize.singular(modelName)]
      return normalizr.denormalize(data, schema, fetchEntities(state))
    }
    
    ${s.export('models')}
  `
}
