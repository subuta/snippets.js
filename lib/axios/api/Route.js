import { build, format, snippets as s, EOL } from 'bld.js'
import pluralize from 'pluralize'
import _ from 'lodash'

import getControllerActions from '../../config/getControllerActions'
import UpdateAction from './actions/Update'
import ShowAction from './actions/Show'
import DestroyAction from './actions/Destroy'
import IndexAction from './actions/All'
import CreateAction from './actions/Create'

export default (RouteConfig, ModelConfig) => {
  const {
    schema
  } = ModelConfig

  const {
    tableName
  } = schema

  const renderAction = (action) => {
    switch (action) {
      case 'index':
        return IndexAction(ModelConfig) + s.EOL
      case 'show':
        return ShowAction(ModelConfig) + s.EOL
      case 'create':
        return CreateAction(ModelConfig) + s.EOL
      case 'update':
        return UpdateAction(ModelConfig) + s.EOL
      case 'destroy':
        return DestroyAction(ModelConfig) + s.EOL
    }
  }

  const keys = getControllerActions(RouteConfig)
  const exports = _.zipObject(keys, keys)

  const actions = _.map(keys, renderAction)

  const model = pluralize.singular(tableName)
  const modelSchema = `${model}`
  const modelListSchema = `${modelSchema}List`

  const imports = s.import([
    ['lodash', '_'],
    ['src/utils/request', 'request'],
    ['normalizr', null, [
      ['normalize']
    ]],
    ['src/utils/schema', null, [
      [modelSchema],
      [modelListSchema],
    ]]
  ])

  return build`
    ${imports}
    
    ${actions}
    
    /* mat Custom action [start] */
    /* mat Custom action [end] */
    
    let actions = ${s.raw(exports)}
    
    /* mat Custom exports [start] */
    /* mat Custom exports [end] */
    
    ${s.export('actions')}
  `
}
