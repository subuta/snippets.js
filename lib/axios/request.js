import { build, format, snippets as s } from 'bld.js'

export default () => {
  const imports = s.import([
    ['axios', 'axios'],
    ['axios/lib/adapters/http', 'httpAdapter'],
    ['lodash', '_'],
    ['file-saver', null, [
      'saveAs'
    ]],
    ['src/views/utils/auth0', 'auth0']
  ])

  return build`
    ${imports}
    
    const isBrowser = typeof window !== 'undefined'

    const origin = isBrowser ? window.location.origin : 'http://localhost:3000'
    
    // switch baseURL
    let baseURL = \`\${origin}/api\`
    
    if (process.env.NODE_ENV === 'development') {
      baseURL = 'http://localhost:3000/api'
    } else if (!isBrowser || process.env.NODE_ENV === 'test') {
      // set httpAdapter while testing.
      baseURL = 'http://localhost:3000/api'
      axios.defaults.adapter = httpAdapter
    }
    
    const request = axios.create({
      baseURL
    })
    
    // Add a request interceptor
    request.interceptors.request.use(
      function(config) {
        const {accessToken} = auth0.getSession()
        // add jwt to header
        config.headers = {
          ...config.headers,
          Authorization: \`Bearer \${accessToken}\`
        }
        return config
      },
      function(error) {
        // Do something with request error
        return Promise.reject(error)
      }
    )
    
    // Add a response interceptor
    request.interceptors.response.use(
      function(response) {
        const {config, data, headers} = response
    
        if (config.responseType === 'blob') {
          const contentDisposition = headers['content-disposition']
          const isAttachment = _.startsWith(
            contentDisposition.toLowerCase(),
            'attachment'
          )
          if (!isAttachment) return response.data
    
          // for file-download
          const fileName =
            _.trim(contentDisposition.split('filename=')[1], '"') || null
          saveAs(data, fileName)
        }
    
        return response.data
      },
      function(error) {
        // Do something with response error
        return Promise.reject(error.response)
      }
    )

    ${s.export('request')}
  `
}
