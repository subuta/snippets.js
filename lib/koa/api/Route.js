import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

import IndexAction from './actions/All'
import ShowAction from './actions/Show'
import CreateAction from './actions/Create'
import UpdateAction from './actions/Update'
import DestroyAction from './actions/Destroy'

import getControllerActions from '../lib/_utils/getControllerActions'

export default (props) => {
  let {
    model,
    config
  } = props

  const {
    imports = []
  } = config

  // Ensure naming convention.
  model = pluralize.singular(model)
  const models = pluralize(model)

  const Import = s.import([
    ...imports,
    ['koa-router', 'Router'],
    ['lodash', '_'],
  ])

  const renderAction = (action) => {
    switch (action) {
      case 'index':
        return IndexAction(props)
      case 'show':
        return ShowAction(props)
      case 'create':
        return CreateAction(props)
      case 'update':
        return UpdateAction(props)
      case 'destroy':
        return DestroyAction(props)
    }
  }

  const actions = _.map(getControllerActions(config), renderAction)

  const prefix = config.prefix || `/${models}`

  return build`
    ${Import}
    
    const ${model} = new Router({
      prefix: ${s.stringify(prefix)}
    })
    
    ${actions}
    
    /* mat Custom actions [start] */
    /* mat Custom actions [end] */
    
    ${s.export(build`
      {
        routes: () => _.cloneDeep(${model}.routes()),
        register: (routers) => {
          /* mat Register [start] */
          /* mat Register [end] */
        }
      }
    `)}
  `
}
