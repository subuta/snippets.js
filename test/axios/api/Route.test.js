import test from 'ava'
import { build, format, snippets as s } from 'bld.js'

import Route from 'lib/axios/api/Route'

import {
  Routes as RoutesConfig,
  Models as ModelsConfig
} from 'test/fixtures/config'

test.beforeEach(async (t) => {

})

test.afterEach((t) => {
})

test('should create axios Route with channel', async (t) => {
  const channelRoute = RoutesConfig['channel']
  const channelModel = ModelsConfig['channel']

  const code = Route(channelRoute, channelModel)

  const expected = build`
    import _ from 'lodash'
    import request from 'src/utils/request'
    import {normalize} from 'normalizr'
    import {channel, channelList} from 'src/utils/schema'
    
    export const index = () => {
      return request.get(\`/channels\`)
    }
    
    export const show = (id) => {
      return request.get(\`/channels/\${id}\`)
    }
    
    export const create = (params) => {
      return request.post(\`/channels\`, {
        channel: params
      })
    }
    
    /* mat Custom action [start] */
    /* mat Custom action [end] */
    
    let actions = {
      index,
      show,
      create
    }
    
    /* mat Custom exports [start] */
    /* mat Custom exports [end] */
    
    export default actions
  `

  t.is(format(code), format(expected))
})

test('should create axios Route with comment', async (t) => {
  const channelRoute = RoutesConfig['comment']
  const channelModel = ModelsConfig['comment']

  const code = Route(channelRoute, channelModel)

  const expected = build`
    import _ from 'lodash'
    import request from 'src/utils/request'
    import {normalize} from 'normalizr'
    import {comment, commentList} from 'src/utils/schema'
    
    export const index = (params) => {
      const {channelId} = params
      return request.get(\`/channels/\$\{channelId\}/comments\`)
    }
    
    export const create = (params) => {
      return request.post(\`/channels/\$\{params.channelId\}/comments\`, {
        comment: params
      })
    }
    
    export const destroy = (id, params) => {
      return request.delete(\`/channels/\$\{params.channelId\}/comments/\$\{id\}\`)
    }
    
    /* mat Custom action [start] */
    /* mat Custom action [end] */
    
    let actions = {
      index,
      create,
      destroy
    }
    
    /* mat Custom exports [start] */
    /* mat Custom exports [end] */
    
    export default actions

  `

  t.is(format(code), format(expected))
})