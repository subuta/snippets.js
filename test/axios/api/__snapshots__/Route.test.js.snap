// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`axios/api/Route should create axios Route with channel 1`] = `
"import _ from 'lodash'
import request from 'src/views/utils/request'
import {normalize} from 'normalizr'
import {channel, channelList} from 'src/views/utils/schema'

export const index = (params) => {
  return request.get(\`/channels\`, {params})
}

export const show = (id) => {
  return request.get(\`/channels/\${id}\`)
}

export const create = (params) => {
  return request.post(\`/channels\`, {
    channel: params
  })
}

/* mat Custom action [start] */
/* mat Custom action [end] */

let actions = {
  index,
  show,
  create
}

/* mat Custom exports [start] */
/* mat Custom exports [end] */

export default actions
"
`;
